File: backend/app/domains/wealth/schemas/sec_funds.py
"""Pydantic schemas for SEC registered fund endpoints."""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field

# ── Registered Funds (N-PORT based) ──────────────────────────────────────


class RegisteredFundSummary(BaseModel):
    """Registered fund row in manager's fund list."""

    cik: str
    fund_name: str
    fund_type: str
    ticker: str | None = None
    total_assets: int | None = None
    last_nport_date: date | None = None
    style_label: str | None = None
    style_confidence: float | None = None


class RegisteredFundListResponse(BaseModel):
    """List of registered funds for a manager."""

    funds: list[RegisteredFundSummary] = Field(default_factory=list)
    total: int = 0


# ── Private Funds (Schedule D) ───────────────────────────────────────────


class PrivateFundSummary(BaseModel):
    """Private fund row from ADV Schedule D."""

    fund_name: str
    fund_type: str | None = None
    gross_asset_value: int | None = None
    investor_count: int | None = None
    is_fund_of_funds: bool | None = None


class PrivateFundListResponse(BaseModel):
    """List of private funds for a manager."""

    funds: list[PrivateFundSummary] = Field(default_factory=list)
    total: int = 0


# ── Fund Detail (Fact Sheet) ────────────────────────────────────────────


class FundFirmInfo(BaseModel):
    """Adviser firm summary embedded in fund detail."""

    crd_number: str
    firm_name: str
    aum_total: int | None = None
    compliance_disclosures: int | None = None
    state: str | None = None
    website: str | None = None


class FundTeamMember(BaseModel):
    """Portfolio manager from ADV Part 2B."""

    person_name: str
    title: str | None = None
    role: str | None = None
    years_experience: int | None = None
    certifications: list[str] = Field(default_factory=list)


class FundStyleInfo(BaseModel):
    """Latest style classification snapshot."""

    style_label: str
    growth_tilt: float
    sector_weights: dict[str, float] = Field(default_factory=dict)
    equity_pct: float | None = None
    fixed_income_pct: float | None = None
    cash_pct: float | None = None
    confidence: float
    report_date: date


class FundDataAvailabilitySchema(BaseModel):
    """Data availability matrix for conditional UI rendering."""

    fund_universe: str
    has_holdings: bool = False
    has_nav_history: bool = False
    has_style_analysis: bool = False
    has_portfolio_manager: bool = False
    has_peer_analysis: bool = False
    disclosure_note: str | None = None


class FundDetailResponse(BaseModel):
    """Full fund detail for fact sheet page."""

    cik: str
    fund_name: str
    fund_type: str
    ticker: str | None = None
    isin: str | None = None
    total_assets: int | None = None
    total_shareholder_accounts: int | None = None
    inception_date: date | None = None
    currency: str = "USD"
    domicile: str = "US"
    last_nport_date: date | None = None

    firm: FundFirmInfo | None = None
    team: list[FundTeamMember] = Field(default_factory=list)
    latest_style: FundStyleInfo | None = None
    data_availability: FundDataAvailabilitySchema


# ── Holdings (N-PORT) ────────────────────────────────────────────────────


class NportHoldingItem(BaseModel):
    """Single N-PORT holding row.

    ``pct_of_nav`` is a pure decimal fraction (0.0741 = 7.41%).
    ``sector`` is the human-readable issuer category label.
    ``issuer_category`` is the raw N-PORT issuerCat code.
    """

    cusip: str | None = None
    isin: str | None = None
    issuer_name: str | None = None
    asset_class: str | None = None
    sector: str | None = None
    issuer_category: str | None = None
    market_value: int | None = None
    quantity: float | None = None
    pct_of_nav: float | None = None
    currency: str | None = None
    fair_value_level: str | None = None


class NportHoldingsPage(BaseModel):
    """Paginated N-PORT holdings response."""

    holdings: list[NportHoldingItem] = Field(default_factory=list)
    available_quarters: list[date] = Field(default_factory=list)
    total_count: int = 0
    total_value: int | None = None


# ── Style History ────────────────────────────────────────────────────────


class StyleSnapshotItem(BaseModel):
    """Single quarter style snapshot."""

    report_date: date
    style_label: str
    growth_tilt: float
    sector_weights: dict[str, float] = Field(default_factory=dict)
    equity_pct: float | None = None
    fixed_income_pct: float | None = None
    cash_pct: float | None = None
    confidence: float


class StyleHistoryResponse(BaseModel):
    """Style classification history with drift detection."""

    snapshots: list[StyleSnapshotItem] = Field(default_factory=list)
    drift_detected: bool = False
    quarters_analyzed: int = 0


# ── Holdings History (sector evolution) ─────────────────────────────────


class TopHoldingItem(BaseModel):
    """Top holding — pct_of_nav is a pure decimal fraction."""

    issuer_name: str | None = None
    sector: str | None = None
    issuer_category: str | None = None
    pct_of_nav: float | None = None
    cusip: str | None = None
    market_value: int | None = None


class HoldingsHistoryResponse(BaseModel):
    """Sector weight evolution across all available N-PORT quarters."""

    quarters: list[str] = Field(default_factory=list)
    sector_series: dict[str, list[float | None]] = Field(default_factory=dict)
    top_holdings_latest: list[TopHoldingItem] = Field(default_factory=list)
    quarters_available: int = 0


# ── Peer Analysis ──────────────────────────────────────────────────────


class PeerFundItem(BaseModel):
    cik: str
    fund_name: str | None = None
    ticker: str | None = None
    expense_ratio_pct: float | None = None
    avg_annual_return_1y: float | None = None
    avg_annual_return_5y: float | None = None
    avg_annual_return_10y: float | None = None


class PeerAnalysisTarget(BaseModel):
    cik: str
    fund_name: str | None = None
    style_label: str | None = None
    expense_ratio_pct: float | None = None
    net_expense_ratio_pct: float | None = None
    avg_annual_return_1y: float | None = None
    avg_annual_return_5y: float | None = None
    avg_annual_return_10y: float | None = None
    portfolio_turnover_pct: float | None = None


class PeerAnalysisResponse(BaseModel):
    """Peer comparison using prospectus stats + style classification."""

    target: PeerAnalysisTarget
    peers: list[PeerFundItem] = Field(default_factory=list)
    peer_count: int = 0
    peer_group: str | None = None
    percentiles: dict[str, int | None] = Field(default_factory=dict)


# ── Reverse Holdings (institutional holders) ───────────────────────────


class InstitutionalHolderItem(BaseModel):
    filer_cik: str
    filer_name: str | None = None
    filer_type: str | None = None
    market_value: int | None = None
    shares: int | None = None
    report_date: str | None = None


class ReverseHoldingsResponse(BaseModel):
    """Institutional holders of a fund via 13F reverse lookup."""

    fund_cusip: str | None = None
    fund_ticker: str | None = None
    holders: list[InstitutionalHolderItem] = Field(default_factory=list)
    total_holders: int = 0
    total_market_value: int = 0
    note: str | None = None


# ── Prospectus Data (SEC RR1) ──────────────────────────────────────────


class ProspectusExpenseExamples(BaseModel):
    y1: int | None = Field(None, alias="1y")
    y3: int | None = Field(None, alias="3y")
    y5: int | None = Field(None, alias="5y")
    y10: int | None = Field(None, alias="10y")

    model_config = {"populate_by_name": True}


class ProspectusFees(BaseModel):
    expense_ratio_pct: float | None = None
    net_expense_ratio_pct: float | None = None
    management_fee_pct: float | None = None
    fee_waiver_pct: float | None = None
    distribution_12b1_pct: float | None = None
    portfolio_turnover_pct: float | None = None
    expense_examples: ProspectusExpenseExamples | None = None


class AnnualReturnItem(BaseModel):
    year: int
    annual_return_pct: float | None = None


class AvgAnnualReturns(BaseModel):
    y1: float | None = Field(None, alias="1y")
    y5: float | None = Field(None, alias="5y")
    y10: float | None = Field(None, alias="10y")

    model_config = {"populate_by_name": True}


class ProspectusDataResponse(BaseModel):
    """SEC prospectus data: fee table + annual return history."""

    series_id: str | None = None
    filing_date: str | None = None
    data_source: str = "SEC DERA RR1 Prospectus"
    fees: ProspectusFees
    annual_returns: list[AnnualReturnItem] = Field(default_factory=list)
    avg_annual_returns: AvgAnnualReturns

File: frontends/wealth/src/lib/types/sec-funds.ts
/** SEC Fund types — registered fund catalog, detail, holdings, style. */

// ── Registered Funds ────────────────────────────────────────────────────

export interface RegisteredFundSummary {
	cik: string;
	fund_name: string;
	fund_type: string;
	ticker: string | null;
	total_assets: number | null;
	last_nport_date: string | null;
	style_label: string | null;
	style_confidence: number | null;
}

export interface RegisteredFundListResponse {
	funds: RegisteredFundSummary[];
	total: number;
}

// ── Private Funds ───────────────────────────────────────────────────────

export interface PrivateFundSummary {
	fund_name: string;
	fund_type: string | null;
	gross_asset_value: number | null;
	investor_count: number | null;
	is_fund_of_funds: boolean | null;
}

export interface PrivateFundListResponse {
	funds: PrivateFundSummary[];
	total: number;
}

// ── Fund Detail ─────────────────────────────────────────────────────────

export interface FundFirmInfo {
	crd_number: string;
	firm_name: string;
	aum_total: number | null;
	compliance_disclosures: number | null;
	state: string | null;
	website: string | null;
}

export interface FundTeamMember {
	person_name: string;
	title: string | null;
	role: string | null;
	years_experience: number | null;
	certifications: string[];
}

export interface FundStyleInfo {
	style_label: string;
	growth_tilt: number;
	sector_weights: Record<string, number>;
	equity_pct: number | null;
	fixed_income_pct: number | null;
	cash_pct: number | null;
	confidence: number;
	report_date: string;
}

export interface FundDataAvailability {
	fund_universe: string;
	has_holdings: boolean;
	has_nav_history: boolean;
	has_style_analysis: boolean;
	has_portfolio_manager: boolean;
	has_peer_analysis: boolean;
	disclosure_note: string | null;
}

export interface FundDetailResponse {
	cik: string;
	fund_name: string;
	fund_type: string;
	ticker: string | null;
	isin: string | null;
	total_assets: number | null;
	total_shareholder_accounts: number | null;
	inception_date: string | null;
	currency: string;
	domicile: string;
	last_nport_date: string | null;
	firm: FundFirmInfo | null;
	team: FundTeamMember[];
	latest_style: FundStyleInfo | null;
	data_availability: FundDataAvailability;
}

// ── Holdings (N-PORT) ───────────────────────────────────────────────────

export interface NportHoldingItem {
	cusip: string | null;
	isin: string | null;
	issuer_name: string | null;
	asset_class: string | null;
	sector: string | null;
	market_value: number | null;
	pct_of_nav: number | null;
	quantity: number | null;
	currency: string | null;
	fair_value_level: string | null;
}

export interface NportHoldingsPage {
	holdings: NportHoldingItem[];
	available_quarters: string[];
	total_count: number;
	total_value: number | null;
}

// ── Style History ───────────────────────────────────────────────────────

export interface StyleSnapshotItem {
	report_date: string;
	style_label: string;
	growth_tilt: number;
	sector_weights: Record<string, number>;
	equity_pct: number | null;
	fixed_income_pct: number | null;
	cash_pct: number | null;
	confidence: number;
}

export interface StyleHistoryResponse {
	snapshots: StyleSnapshotItem[];
	drift_detected: boolean;
	quarters_analyzed: number;
}

// ── Defaults ────────────────────────────────────────────────────────────

export const EMPTY_REGISTERED_FUNDS: RegisteredFundListResponse = { funds: [], total: 0 };
export const EMPTY_PRIVATE_FUNDS: PrivateFundListResponse = { funds: [], total: 0 };
export const EMPTY_HOLDINGS: NportHoldingsPage = {
	holdings: [],
	available_quarters: [],
	total_count: 0,
	total_value: null,
};
export const EMPTY_STYLE_HISTORY: StyleHistoryResponse = {
	snapshots: [],
	drift_detected: false,
	quarters_analyzed: 0,
};
